### PR TITLE
PCHR-2208: Default value for entitlements

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
My leave page on SSP doesn't load when user doesn't have any entitlement

## Before
Users with no entitlement got uncaught error in`/my-leave#/my-leave/report`page and the UI froze.
Also 'Record new absence' - modal doesn't load

## After

1. The My Leave report loads
2. The my leave report page hides the absence types which has 0 entitlement, does not allow negative balance and does not allow accrual requests.
3. Displays "You don't have any leave entitlement" in place of absence type dropdown when all the absence types have 0 entitlement.

## Technical Details
1. In My Leave Report Page, a new scope variable `absenceTypesFiltered` is created which is used to show only the following absence types
```
!((absenceType.entitlement === 0) && (absenceType.allow_overuse !== '1') &&
 (absenceType.allow_accruals_request !== '1'))
```
2. `absence-type-api` only loads the active absence types by default, but it can be overridden by passing a parameter.

3. `request-ctrl` throws a `No entitlement` error, and in such cases the error is not shown to the user in the **red error band**.

4. When opening the Leave Request Popup, in no absence types can be found(because it is already filtered), the UI is blocked and `You don't have any leave entitlement` message is shown to the user.

5. `chr_leave-request-modal__footer` is added to the footer of the leave request popup, it is used to fix styling bug in https://github.com/compucorp/civihr-employee-portal-theme/pull/161
---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html [hunk 1]
```diff
@@ -34,18 +34,19 @@ <h4 class="modal-title">
       </ui-select>
     </div>
     <div
-      ng-class="{'chr_disabled': !$ctrl.request.contact_id}"
+      ng-class="{'chr_disabled': (!$ctrl.request.contact_id || $ctrl.noEntitlement)}"
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html]: can the entire dialog be disabled once instead of at multiple places?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/leave-request-ctrl_test.js [hunk 2]
```diff
@@ -1197,40 +1198,56 @@
...
+
+              it('is not available', function () {
+                expect(availableStatuses).not.toContain(cancelStatus);
+              });
+            });
+          });
+
+          describe('when entitlement is not present', function() {
+            beforeEach(function() {
+              spyOn(Entitlement, 'all').and.callFake(function () {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/sub-controllers/leave-request-ctrl_test.js]: can the spyOn be on EntitlementAPI itself and avoid adding Entitlement in this file?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 3]
```diff
@@ -154,25 +154,29 @@
...
               });
 
-              it('has stored the entitlement value in each absence type', function () {
+              it('has stored the 0 value for entitlement, remainder for absence types which does not have entitlement', function () {
                 _.forEach(controller.absenceTypes, function (absenceType) {
-                  var value = absenceType.entitlement;
-
-                  expect(value).toBeDefined();
-                  expect(value).toEqual(_.find(controller.entitlements, function (entitlement) {
+                  var entitlement = _.find(controller.entitlements, function (entitlement) {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: can these 2 tests be in different describe with setup of test data happening in beforeEach instead of it?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 4]
```diff
@@ -657,7 +658,10 @@ define([
 
               initialCommentsLength = self.request.comments.length;
             }
-          });
+          })
+          .catch(function (error) {
+            self.noEntitlement = ( error === 'No entitlement' );
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Can this message be more descriptive like "You do not have entitlements, kindly get in touch with your manager to get some" etc?

**Comment by mi-kon** [comment]
@deb1990 instead of showing the text as it is now, please still display the select dropdown, make it disabled and show "You don't have leave entitlement" message inside the disabled dropdown.

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html (hunk 1 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html]: Code has changed to `$ctrl.absenceTypes.length === 0` but I am disabling only once.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js (hunk 3 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: Did not do that because the test case is inside a forEach, so if I create 2 test cases the loop will happen twice. Also beforeEach wont help as the `expect`is inside the loop.

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html [hunk 5]
```diff
@@ -33,17 +33,23 @@ <h4 class="modal-title">
...
             <select
+              ng-hide="$ctrl.absenceTypes.length === 0"
               ng-disabled="$ctrl.isMode('view') || ($ctrl.isRole('manager') && !$ctrl.isMode('create'))"
               name="absenceTypeSelect"
               ng-options="absenceType.id as absenceType.title for absenceType in $ctrl.absenceTypes"
               ng-model="$ctrl.request.type_id"
               ng-change="$ctrl.updateBalance()"></select>
+            <select
+              ng-show="$ctrl.absenceTypes.length === 0"
+              disabled="disabled">
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html]: why not use ng-disabled like above?

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html [hunk 6]
```diff
@@ -364,7 +370,7 @@ <h4 class="modal-title">
     <div ng-repeat="errMsg in $ctrl.errors track by $index">{{errMsg}}</div>
   </div>
   <crm-loading show="!$ctrl.submitting">
-    <div class="modal-footer">
+    <div class="modal-footer chr_leave-request-modal__footer">
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html]: why is this added? ps add to pr description

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 7]
```diff
@@ -29,6 +29,7 @@ define([
       var mode = ''; // can be edit, create, view
       var role = '';
       var initialCommentsLength = 0; // number of comments when the request model is loaded
+      var NO_ENTITLEMENT_ERROR = 'No entitlement';
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: As this variable stores a constant string value, its made in all caps. which is a common coding practise.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/data/entitlement-data.js [hunk 8]
```diff
@@ -66,54 +66,57 @@ define([
...
   /**
    * Generates an entitlement for each absence type in each period
    *
    * @param {Boolean} withRemainder if the remainder data should be attached
    * @return {Array}
    */
-  function generateEntitlements(withRemainder) {
-    var i = 1, values = [];
+  function generateEntitlements (withRemainder) {
+    var i = 1;
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/data/entitlement-data.js]: `i` is used for iterator, it's a common coding convention. Also there was a discussion related to this and we all agreed that `i` is fine.

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html (hunk 5 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup.html]: This select box is used to only show the error message, so it is always disabled and does not depend on a condition. Its visibility is dependent on a condition, which is specified.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 9]
```diff
@@ -1103,6 +1108,9 @@ define([
           .then(function (entitlements) {
             // create a list of absence types with a `balance` property
             self.absenceTypes = mapAbsenceTypesWithBalance(absenceTypesAndIds.types, entitlements);
+            if (!self.absenceTypes.length) {
+              return $q.reject(NO_ENTITLEMENT_ERROR);
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: It is not mandatory, as in if I dont mention anything it will resolve the promise by default. That is how it was working before, I just added condition to reject the promise in certain condition.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 10]
```diff
@@ -666,7 +667,11 @@ define([
 
             self.postContactSelection = false;
           })
-          .catch(handleError.bind(self));
+          .catch(function (error) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: It is handled in the parent function(from where `initAfterContactSelection` is called) i.e. `_init`.
```javascript
then(function () {
  return self.initAfterContactSelection();
})
.catch(handleError.bind(self));
```
